File: apps/opik-backend/src/main/java/com/comet/opik/api/resources/v1/events/tools/TraceCompressor.java
package com.comet.opik.api.resources.v1.events.tools;

import com.comet.opik.api.Span;
import com.comet.opik.api.Trace;
import com.comet.opik.api.attachment.AttachmentInfo;
import com.comet.opik.utils.JsonUtils;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import jakarta.inject.Singleton;
import lombok.NonNull;

import java.util.List;
import java.util.Map;
import java.util.UUID;

/**
 * Bespoke adaptive compressor for traces. The cached FULL form is the
 * composite {@code {trace, spans}}; this compressor produces three tiers:
 *
 * <ul>
 *   <li>FULL — &lt; {@link #FULL_TOKEN_LIMIT}: full composite untouched.</li>
 *   <li>MEDIUM — between {@link #FULL_TOKEN_LIMIT} and {@link #MEDIUM_TOKEN_LIMIT}:
 *       composite with all string values longer than {@link #STRING_TRUNCATION_LENGTH}
 *       replaced via the path-aware truncator. jq paths target the cached
 *       composite (e.g. {@code .trace.input}, {@code .spans[2].output}).</li>
 *   <li>SKELETON — ≥ {@link #MEDIUM_TOKEN_LIMIT}: identity fields + counts +
 *       a minimal span tree (id / name / type / parent / children only).</li>
 * </ul>
 *
 * <p>SUMMARY isn't in this compressor's ladder and collapses to SKELETON.
 *
 * <p>Returned {@link CompressionResult#payload()} is the bare inner content;
 * {@code ReadTool} wraps it in the {@code data} field of the response envelope.
 */
@Singleton
public final class TraceCompressor implements EntityCompressor {

    static final int FULL_TOKEN_LIMIT = 8_000;
    static final int MEDIUM_TOKEN_LIMIT = 50_000;
    static final int STRING_TRUNCATION_LENGTH = 1000;

    @Override
    public EntityType type() {
        return EntityType.TRACE;
    }

    /**
     * Builds the FULL composite JSON {@code {"trace": ..., "spans": [...]}}. Exposed
     * so {@code ReadTool} can cache it before invoking compression.
     */
    public JsonNode buildFullJson(@NonNull Trace trace, @NonNull List<Span> spans) {
        var mapper = JsonUtils.getMapper();
        ObjectNode node = mapper.createObjectNode();
        node.set("trace", mapper.valueToTree(trace));
        node.set("spans", mapper.valueToTree(spans));
        return node;
    }

    public CompressionResult compress(@NonNull JsonNode fullJson,
            @NonNull Trace trace,
            @NonNull List<Span> spans,
            CompressionTier forcedTier) {
        // Delegate to the attachments overload (which owns the WITH_JQ_HINT default) rather than
        // re-specifying the suffix here, so the default lives in exactly one place.
        return compress(fullJson, trace, spans, forcedTier, Map.of(), List.of());
    }

    /**
     * Variant that lets the caller pick the truncation suffix style. Pass
     * {@link PathAwareTruncator.SuffixStyle#BARE} when the cache itself was
     * capped, since the {@code use jq(...) to see full} pointer would be a
     * lie — the cache has no full value to recover.
     */
    CompressionResult compress(@NonNull JsonNode fullJson,
            @NonNull Trace trace,
            @NonNull List<Span> spans,
            CompressionTier forcedTier,
            @NonNull PathAwareTruncator.SuffixStyle suffix) {
        return compress(fullJson, trace, spans, forcedTier, suffix, Map.of(), List.of());
    }

    /**
     * Variant that enriches <em>every</em> tier with the supplied attachment summaries — per-span
     * attachments (matched by span id) and trace-level attachments — so a caller injecting the
     * result into a prompt (the {@code {{trace}}} variable) gets the real {@code file_name}s the
     * judge needs for {@code get_attachment} on any of the trace's spans, not just the trace itself.
     * Attachments are metadata only; when both maps are empty the payload is identical to the
     * attachment-free path, so callers that don't supply them (e.g. {@code ReadTool}, which surfaces
     * attachments separately) pay nothing. Stays id-agnostic: the caller adds any id envelope on top.
     */
    public CompressionResult compress(@NonNull JsonNode fullJson,
            @NonNull Trace trace,
            @NonNull List<Span> spans,
            CompressionTier forcedTier,
            Map<UUID, List<AttachmentInfo>> spanAttachments,
            List<AttachmentInfo> traceAttachments) {
        return compress(fullJson, trace, spans, forcedTier, PathAwareTruncator.SuffixStyle.WITH_JQ_HINT,
                spanAttachments, traceAttachments);
    }

    CompressionResult compress(@NonNull JsonNode fullJson,
            @NonNull Trace trace,
            @NonNull List<Span> spans,
            CompressionTier forcedTier,
            @NonNull PathAwareTruncator.SuffixStyle suffix,
            Map<UUID, List<AttachmentInfo>> spanAttachments,
            List<AttachmentInfo> traceAttachments) {

        Map<UUID, List<AttachmentInfo>> spanAtt = spanAttachments != null ? spanAttachments : Map.of();
        List<AttachmentInfo> traceAtt = traceAttachments != null ? traceAttachments : List.of();
        boolean hasAttachments = !spanAtt.isEmpty() || !traceAtt.isEmpty();

        CompressionTier tier = pickTier(fullJson, forcedTier);
        JsonNode payload = switch (tier) {
            case FULL -> hasAttachments ? enrichComposite(fullJson.deepCopy(), spanAtt, traceAtt) : fullJson;
            case MEDIUM -> {
                JsonNode truncated = PathAwareTruncator.truncate(fullJson, STRING_TRUNCATION_LENGTH, suffix);
                yield hasAttachments ? enrichComposite(truncated, spanAtt, traceAtt) : truncated;
            }
            case SKELETON, SUMMARY -> buildSkeleton(trace, spans, spanAtt, traceAtt);
        };
        CompressionTier reportedTier = tier == CompressionTier.SUMMARY ? CompressionTier.SKELETON : tier;
        return CompressionResult.builder()
                .payload(payload)
                .tier(reportedTier)
                .build();
    }

    private static CompressionTier pickTier(JsonNode fullJson, CompressionTier forced) {
        if (forced != null) {
            return forced;
        }
        int estimate = Tokens.estimate(fullJson.toString());
        if (estimate < FULL_TOKEN_LIMIT) {
            return CompressionTier.FULL;
        }
        if (estimate < MEDIUM_TOKEN_LIMIT) {
            return CompressionTier.MEDIUM;
        }
        return CompressionTier.SKELETON;
    }

    /**
     * Adds attachment summaries to a {@code {trace, spans}} composite (FULL / MEDIUM tiers) in
     * place: trace-level attachments on the {@code trace} node, per-span attachments on each
     * {@code spans[]} entry matched by its {@code id}. The node must already be safe to mutate
     * (a deep copy of the cache, or a freshly truncated tree).
     */
    private static JsonNode enrichComposite(JsonNode composite,
            Map<UUID, List<AttachmentInfo>> spanAttachments, List<AttachmentInfo> traceAttachments) {
        if (composite instanceof ObjectNode root) {
            if (root.get("trace") instanceof ObjectNode traceNode) {
                setAttachments(traceNode, traceAttachments);
            }
            if (root.get("spans") instanceof ArrayNode spansArray) {
                for (JsonNode spanNode : spansArray) {
                    if (spanNode instanceof ObjectNode obj && obj.hasNonNull("id")) {
                        setAttachments(obj, spanAttachments.get(UUID.fromString(obj.get("id").asText())));
                    }
                }
            }
        }
        return composite;
    }

    private static ObjectNode buildSkeleton(Trace trace, List<Span> spans,
            Map<UUID, List<AttachmentInfo>> spanAttachments, List<AttachmentInfo> traceAttachments) {
        var mapper = JsonUtils.getMapper();
        ObjectNode node = mapper.createObjectNode();
        node.put("name", trace.name());
        if (trace.projectId() != null) {
            node.put("project_id", trace.projectId().toString());
        }
        node.put("span_count", spans.size());
        node.put("error_count", (int) spans.stream().filter(s -> s.errorInfo() != null).count());
        if (trace.duration() != null) {
            node.put("total_duration_ms", trace.duration());
        }
        node.set("span_tree", SpanHierarchy.toTree(spans, span -> buildSkeletonNode(span, spanAttachments)));
        setAttachments(node, traceAttachments);
        return node;
    }

    private static ObjectNode buildSkeletonNode(Span span, Map<UUID, List<AttachmentInfo>> spanAttachments) {
        ObjectNode node = JsonUtils.getMapper().createObjectNode();
        node.put("id", span.id().toString());
        node.put("name", span.name());
        node.put("type", span.type() != null ? span.type().toString() : null);
        setAttachments(node, spanAttachments.get(span.id()));
        return node;
    }

    private static void setAttachments(ObjectNode node, List<AttachmentInfo> attachments) {
        if (attachments == null || attachments.isEmpty()) {
            return;
        }
        ArrayNode array = AttachmentSummaries.toJsonArray(attachments);
        if (!array.isEmpty()) {
            node.set("attachments", array);
        }
    }
}

File: sdks/python/src/opik/evaluation/suite_evaluators/agentic/compression/trace_compressor.py
"""Bespoke adaptive compressor for traces.

Three tiers, picked by token estimate against the FULL composite:
- FULL — full `{"trace": ..., "spans": [...]}` untouched.
- MEDIUM — same composite, but every string longer than
  `STRING_TRUNCATION_LENGTH` chars replaced via path-aware truncator.
- SKELETON — identity fields + counts + a minimal span tree (id /
  name / type, with parent/child structure).

Backend parity: thresholds and skeleton shape match
`TraceCompressor.java`. Suffix wording diverges (we point at `scan`,
not `jq`) — see design doc §3.5.
"""

import datetime
import logging
from typing import Any, Dict, List, Optional

from . import path_aware_truncator, tier as tier_module, tokens

LOGGER = logging.getLogger(__name__)

FULL_TOKEN_LIMIT = 8_000
MEDIUM_TOKEN_LIMIT = 50_000
STRING_TRUNCATION_LENGTH = 1_000


def build_full_json(
    trace: Dict[str, Any], spans: List[Dict[str, Any]]
) -> Dict[str, Any]:
    """Compose the FULL view `{"trace": trace, "spans": spans}`.

    Caller passes already-serialized dicts (see `context._serialize_trace`
    / `_serialize_span`); we don't reach back into the typed models, so
    this stays pure.
    """
    return {"trace": trace, "spans": spans}


def compress(
    full_json: Dict[str, Any],
    trace: Dict[str, Any],
    spans: List[Dict[str, Any]],
    parent_by_child: Dict[str, Optional[str]],
    forced_tier: Optional[tier_module.CompressionTier] = None,
) -> tier_module.CompressionResult:
    """Render `full_json` at the appropriate tier and return the result.

    `forced_tier` lets the caller pin the tier (used by `read(tier=...)`);
    when None, the tier is picked from the token estimate. `parent_by_child`
    is only consulted at SKELETON tier to build the nested span tree.
    """
    tier = _pick_tier(full_json, forced_tier)
    if tier is tier_module.CompressionTier.FULL:
        payload: Dict[str, Any] = full_json
    elif tier is tier_module.CompressionTier.MEDIUM:
        payload = path_aware_truncator.truncate_strings(
            full_json, STRING_TRUNCATION_LENGTH
        )
    else:
        # SKELETON / SUMMARY collapse to the skeleton view; the reported
        # tier flattens SUMMARY → SKELETON since this compressor doesn't
        # ship a separate SUMMARY rendering.
        payload = _build_skeleton(trace, spans, parent_by_child)

    reported_tier = (
        tier_module.CompressionTier.SKELETON
        if tier is tier_module.CompressionTier.SUMMARY
        else tier
    )
    return tier_module.CompressionResult(payload=payload, tier=reported_tier)


def _pick_tier(
    full_json: Dict[str, Any], forced: Optional[tier_module.CompressionTier]
) -> tier_module.CompressionTier:
    if forced is not None:
        return forced
    estimate = tokens.estimate_tokens(full_json)
    if estimate < FULL_TOKEN_LIMIT:
        return tier_module.CompressionTier.FULL
    if estimate < MEDIUM_TOKEN_LIMIT:
        return tier_module.CompressionTier.MEDIUM
    return tier_module.CompressionTier.SKELETON


def _build_skeleton(
    trace: Dict[str, Any],
    spans: List[Dict[str, Any]],
    parent_by_child: Dict[str, Optional[str]],
) -> Dict[str, Any]:
    error_count = sum(1 for span in spans if span.get("error_info") is not None)
    return {
        "name": trace.get("name"),
        "project_name": trace.get("project_name"),
        "span_count": len(spans),
        "error_count": error_count,
        "total_duration_ms": _duration_ms(
            trace.get("start_time"), trace.get("end_time")
        ),
        "span_tree": _build_span_tree(spans, parent_by_child),
    }


def _build_span_tree(
    spans: List[Dict[str, Any]], parent_by_child: Dict[str, Optional[str]]
) -> List[Dict[str, Any]]:
    """Build a nested tree of `{id, name, type, spans: [...]}` nodes.

    Orphan spans (parent id present but parent span not in the list) are
    promoted to roots so they remain visible — same fallback the backend
    `SpanHierarchy.toTree` applies.
    """
    nodes: Dict[str, Dict[str, Any]] = {
        span["id"]: {
            "id": span["id"],
            "name": span.get("name"),
            "type": span.get("type"),
            "spans": [],
        }
        for span in spans
    }
    roots: List[Dict[str, Any]] = []
    for span in spans:
        span_id = span["id"]
        parent_id = parent_by_child.get(span_id)
        if parent_id is None or parent_id not in nodes:
            roots.append(nodes[span_id])
        else:
            nodes[parent_id]["spans"].append(nodes[span_id])
    return roots


def _duration_ms(start: Optional[str], end: Optional[str]) -> Optional[float]:
    if start is None or end is None:
        return None
    try:
        start_dt = datetime.datetime.fromisoformat(start)
        end_dt = datetime.datetime.fromisoformat(end)
    except (TypeError, ValueError):
        # Best-effort: skeleton-tier duration is informational, not load-bearing.
        LOGGER.info(
            "Failed to parse trace timestamps for duration_ms "
            "(start=%r, end=%r); returning None.",
            start,
            end,
            exc_info=True,
        )
        return None
    return (end_dt - start_dt).total_seconds() * 1000.0
